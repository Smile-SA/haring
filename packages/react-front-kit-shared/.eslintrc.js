module.exports = {
  extends: ['custom'],
  parserOptions: {
    project: ['./tsconfig.json'],
    tsconfigRootDir: __dirname,
  },
  root: true,
};
