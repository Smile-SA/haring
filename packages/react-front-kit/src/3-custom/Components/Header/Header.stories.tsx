import type { Meta, StoryObj } from '@storybook/react';

import { Avatar, Menu } from '@mantine/core';
import { expect } from '@storybook/jest';
import { userEvent, within } from '@storybook/testing-library';

import { primaryTheme } from '../../../theme';
import { DropdownButton } from '../DropdownButton/DropdownButton';

import { Header as Cmp } from './Header';

const meta = {
  argTypes: {
    children: {
      table: {
        disable: true,
      },
    },
    left: {
      table: {
        disable: true,
      },
    },
    right: {
      table: {
        disable: true,
      },
    },
    searchTheme: {
      table: {
        disable: true,
      },
    },
  },
  component: Cmp,
  parameters: {
    layout: 'fullscreen',
  },
  tags: ['autodocs'],
  title: '3-custom/Components/Header',
} satisfies Meta<typeof Cmp>;

export default meta;
type IStory = StoryObj<typeof meta>;

export const Header: IStory = {
  args: {
    children: (
      <>
        <a href="#">Espace documentaire</a>
        <a href="#">Espace workflow</a>
        <a href="#">Archives</a>
      </>
    ),
    childrenComponent: 'nav',
    height: 90,
    left: <img alt="logo" height="58" src="./logo.svg" width="128" />,
    right: (
      <>
        <DropdownButton label="Mon espace">
          <Menu.Item component="a" href="#">
            Calico
          </Menu.Item>
          <Menu.Item component="a" href="#">
            Espace RH
          </Menu.Item>
          <Menu.Item component="a" href="#">
            Aventure IA
          </Menu.Item>
          <Menu.Item component="a" href="#">
            Lunette & CO
          </Menu.Item>
        </DropdownButton>
        <Avatar src="https://images.unsplash.com/photo-1535713875002-d1d0cf377fde?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=250&q=80" />
      </>
    ),
    searchTheme: primaryTheme,
    withBorder: false,
  },
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement);
    await userEvent.click(canvas.getByTestId('search'));
    await expect(canvas.getByTestId('searchBar')).toBeInTheDocument();
  },
};
