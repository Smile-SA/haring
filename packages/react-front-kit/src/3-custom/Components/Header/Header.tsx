import type { HeaderProps, MantineThemeOverride } from '@mantine/core';
import type {
  ChangeEventHandler,
  ElementType,
  FormEvent,
  MouseEventHandler,
  ReactNode,
} from 'react';

import {
  Button,
  Flex,
  Header as MantineHeader,
  MantineProvider,
  createStyles,
  useMantineTheme,
} from '@mantine/core';
import { useClickOutside } from '@mantine/hooks';
import { MagnifyingGlass } from '@phosphor-icons/react';
import { useState } from 'react';

import { HeaderSearch } from '../HeaderSearch/HeaderSearch';

const useStyles = createStyles((theme) => ({
  around: {
    alignItems: 'center',
    gap: theme.spacing.xs,
  },
  button: {
    background: 'transparent',
    borderRadius: 0,
    position: 'relative',
  },
  buttonOpened: {
    '&::after': {
      background:
        theme.colorScheme === 'light'
          ? theme.colors.gray[3]
          : theme.colors.gray[8],
      content: '""',
      display: 'block',
      height: 36,
      position: 'absolute',
      right: 0,
      top: '50%',
      translate: '0 -50%',
      width: 1,
    },
    background: theme.fn.primaryColor(),
    color: theme.white,
  },
  container: {
    alignItems: 'center',
    justifyContent: 'space-between',
    padding: '16px 64px',
    position: 'relative',
    width: '100%',
  },
  menu: {
    alignItems: 'center',
    gap: theme.spacing.xl,
    left: '50%',
    margin: 'auto',
    position: 'absolute',
    top: '50%',
    translate: '-50% -50%',
  },
}));

interface IHeaderProps extends Omit<HeaderProps, 'height' | 'left' | 'right'> {
  childrenComponent?: ElementType;
  height?: number;
  left?: ReactNode;
  onSearchChange?: ChangeEventHandler<HTMLInputElement>;
  onSearchClear?: MouseEventHandler<HTMLButtonElement>;
  onSearchSubmit?: (event: FormEvent) => void;
  right?: ReactNode;
  searchTheme?: MantineThemeOverride;
  searchValue?: string;
}

export function Header(props: IHeaderProps): JSX.Element {
  const {
    children,
    childrenComponent,
    height = 90,
    left,
    onSearchChange,
    onSearchClear,
    onSearchSubmit,
    right,
    searchTheme,
    searchValue,
    ...headerProps
  } = props;
  const { classes } = useStyles();
  const [opened, setOpened] = useState(false);
  const header = useClickOutside(() => setOpened(false));
  const defaultTheme = useMantineTheme();

  function handleClick(): void {
    setOpened?.(!opened);
  }

  const buttonClasses = [classes.button];
  if (opened) {
    buttonClasses.push(classes.buttonOpened);
  }

  return (
    <MantineHeader
      ref={header}
      height={opened ? height + 110 : height}
      {...headerProps}
    >
      <Flex className={classes.container} h={height}>
        <Flex className={classes.around}>{left}</Flex>
        <Flex
          className={classes.menu}
          // @ts-expect-error wrong type for Flex component
          component={childrenComponent}
        >
          {children}
        </Flex>
        <Flex className={classes.around}>
          <Button
            className={buttonClasses.join(' ')}
            h={height}
            onClick={handleClick}
            variant="white"
            w={height}
          >
            <MagnifyingGlass size={32} />
          </Button>
          {Boolean(opened) && (
            <MantineProvider theme={searchTheme ?? defaultTheme}>
              <HeaderSearch
                onChange={onSearchChange}
                onClear={onSearchClear}
                onSubmit={onSearchSubmit}
                opened={opened}
                value={searchValue}
              />
            </MantineProvider>
          )}
          {right}
        </Flex>
      </Flex>
    </MantineHeader>
  );
}

Header.defaultProps = {
  withBorder: false,
};
