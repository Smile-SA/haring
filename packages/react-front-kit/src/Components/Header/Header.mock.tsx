import type { IHeaderNavMenu } from '../HeaderNav/HeaderNav';
import type { ReactElement } from 'react';

import { Avatar, Menu } from '@mantine/core';

import { DropdownButton } from '../DropdownButton/DropdownButton';
import { HeaderNav } from '../HeaderNav/HeaderNav';

export const menusMock: IHeaderNavMenu<number>[] = [
  {
    children: [
      { id: 4, label: 'Releases', url: '#' },
      { id: 5, label: 'Account', url: '#' },
      {
        children: [
          { id: 8, label: 'Wiki pages', url: '#' },
          { id: 9, label: 'Settings', url: '#' },
        ],
        id: 6,
        label: 'Upcoming releases',
        url: '#',
      },
      { id: 7, label: 'Lorem ipsum dolor', url: '#' },
    ],
    id: 1,
    label: 'Espace documentaire',
    url: '#',
  },
  { id: 2, label: 'Espace workflow', url: '#' },
  { id: 3, label: 'Archives', url: '#' },
];

export const childrenMock = (isMobile: boolean): ReactElement => {
  return <HeaderNav isMobile={isMobile} menus={menusMock} />;
};

export const leftContentMock = (
  <img alt="logo" height="58" src="./logo.svg" width="128" />
);

export const rightContentMobileMock = (
  <DropdownButton label="Mon espace">
    <Menu.Item component="a" href="#">
      Calico
    </Menu.Item>
    <Menu.Item component="a" href="#">
      Espace RH
    </Menu.Item>
    <Menu.Item component="a" href="#">
      Aventure IA
    </Menu.Item>
    <Menu.Item component="a" href="#">
      Lunette & CO
    </Menu.Item>
  </DropdownButton>
);

export const rightContentMock = (
  <>
    {rightContentMobileMock}
    <Avatar src="https://images.unsplash.com/photo-1535713875002-d1d0cf377fde?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=250&q=80" />
  </>
);
