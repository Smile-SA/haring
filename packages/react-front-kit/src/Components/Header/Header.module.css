.around {
  align-items: center;
  gap: var(--mantine-spacing-xs);
}

.button {
  &::after {
    content: '';
    display: block;
    height: 36px;
    position: absolute;
    right: 0;
    top: 50%;
    translate: 0 -50%;
    width: 1px;

    /*noinspection CssInvalidAtRule*/
    @mixin light {
      background: var(--mantine-color-gray-3);
    }
    /*noinspection CssInvalidAtRule*/
    @mixin dark {
      background: var(--mantine-color-gray-8);
    }
  }

  background: transparent;
  border-radius: 0;
  position: relative;
}

.buttonOpened {
  &::after {
    display: none;
  }

  background: var(--mantine-primary-color-filled);
  color: var(--mantine-color-white);
}

.container {
  align-items: center;
  justify-content: space-between;
  padding: 16px 64px;
  position: relative;
  width: 100%;
}

.menu {
  align-items: center;
  gap: var(--mantine-spacing-xl);
  left: 50%;
  margin: auto;
  position: absolute;
  top: 50%;
  translate: -50% -50%;
}

.none {
  display: none;
}

.sizeDesktop {
  display: initial;

  /*noinspection CssInvalidAtRule*/
  @mixin smaller-than $mantine-breakpoint-sm {
    display: none;
  }
}

.sizeMobile {
  display: none;

  /*noinspection CssInvalidAtRule*/
  @mixin smaller-than $mantine-breakpoint-sm {
    display: initial;
  }
}
