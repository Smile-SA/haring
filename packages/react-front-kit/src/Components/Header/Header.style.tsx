import { createStyles } from '@mantine/core';

export const useStyles = createStyles((theme) => ({
  around: {
    alignItems: 'center',
    gap: theme.spacing.xs,
  },
  button: {
    '&::after': {
      background:
        theme.colorScheme === 'light'
          ? theme.colors.gray[3]
          : theme.colors.gray[8],
      content: '""',
      display: 'block',
      height: 36,
      position: 'absolute',
      right: 0,
      top: '50%',
      translate: '0 -50%',
      width: 1,
    },
    background: 'transparent',
    borderRadius: 0,
    position: 'relative',
  },
  buttonOpened: {
    '&::after': {
      display: 'none',
    },
    background: theme.fn.primaryColor(),
    color: theme.white,
  },
  container: {
    alignItems: 'center',
    justifyContent: 'space-between',
    padding: '16px 64px',
    position: 'relative',
    width: '100%',
  },
  menu: {
    alignItems: 'center',
    gap: theme.spacing.xl,
    left: '50%',
    margin: 'auto',
    position: 'absolute',
    top: '50%',
    translate: '-50% -50%',
  },
  none: {
    display: 'none',
  },
  sizeDesktop: {
    display: 'initial',
    [theme.fn.smallerThan('md')]: {
      display: 'none',
    },
  },
  sizeMobile: {
    display: 'none',
    [theme.fn.smallerThan('md')]: {
      display: 'initial',
    },
  },
}));
