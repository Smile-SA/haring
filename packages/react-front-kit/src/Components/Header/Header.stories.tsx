import type { Meta, StoryObj } from '@storybook/react';

import { primaryTheme } from '@smile/react-front-kit-shared';
import { useStorybookArgsConnect } from '@smile/react-front-kit-shared/storybook-utils';
import { action } from '@storybook/addon-actions';
import { expect } from '@storybook/jest';
import { userEvent, within } from '@storybook/testing-library';

import { Header as Cmp } from './Header';
import {
  childrenMock,
  leftContentMock,
  rightContentMobileMock,
  rightContentMock,
} from './Header.mock';

const meta = {
  argTypes: {
    children: {
      control: false,
    },
    childrenComponent: {
      control: 'text',
    },
    left: {
      control: false,
    },
    right: {
      control: false,
    },
  },
  component: Cmp,
  decorators: [
    function Component(Story, ctx) {
      const args = useStorybookArgsConnect(ctx.args, {
        onSearchChange: 'searchValue',
      });
      return <Story args={{ ...args }} />;
    },
  ],
  parameters: {
    layout: 'fullscreen',
  },
  tags: ['autodocs'],
  title: '3-custom/Components/Header',
} satisfies Meta<typeof Cmp>;

export default meta;
type IStory = StoryObj<typeof meta>;

export const Header: IStory = {
  args: {
    children: childrenMock(false),
    childrenComponent: 'nav',
    hasResponsiveMode: true,
    height: 90,
    left: leftContentMock,
    mobileProps: {
      children: childrenMock(true),
      right: rightContentMobileMock,
    },
    onSearchSubmit: action('search input submitted'),
    right: rightContentMock,
    searchTheme: primaryTheme,
    searchValue: '',
    withBorder: false,
  },
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement);
    await userEvent.click(canvas.getByTestId('search'));
    await expect(canvas.getByTestId('searchBar')).toBeInTheDocument();
  },
};
