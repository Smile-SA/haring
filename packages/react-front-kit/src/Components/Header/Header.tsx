'use client';

import type { HeaderProps, MantineThemeOverride } from '@mantine/core';
import type { ElementType, FormEvent, ReactElement, ReactNode } from 'react';

import {
  Button,
  Flex,
  Header as MantineHeader,
  MantineProvider,
  useMantineTheme,
} from '@mantine/core';
import { useClickOutside } from '@mantine/hooks';
import { MagnifyingGlass } from '@phosphor-icons/react';
import { useState } from 'react';

import { HeaderMobile } from '../HeaderMobile/HeaderMobile';
import { HeaderSearch } from '../HeaderSearch/HeaderSearch';

import { useStyles } from './Header.style';

export interface IHeaderProps
  extends Omit<HeaderProps, 'height' | 'left' | 'right'> {
  children: ReactNode;
  childrenComponent?: ElementType;
  hasResponsiveMode?: boolean;
  hasSearch?: boolean;
  height?: number;
  left?: ReactNode;
  mobileProps?: {
    children?: ReactNode;
    height?: number;
    left?: ReactNode;
    menuAriaLabel?: string;
    right?: ReactNode;
    searchPlaceholder?: string;
  };
  onSearchChange?: (value: string) => void;
  onSearchSubmit?: (event: FormEvent) => void;
  right?: ReactNode;
  searchAriaLabel?: string;
  searchPlaceholder?: string;
  searchTheme?: MantineThemeOverride;
  searchValue?: string;
}

/** Additional props will be forwarded to the [Mantine AppShell (Header) component](https://mantine.dev/core/app-shell/) */
export function Header(props: IHeaderProps): ReactElement {
  const {
    children,
    childrenComponent,
    hasResponsiveMode = true,
    hasSearch = true,
    height = 90,
    left,
    onSearchChange,
    onSearchSubmit,
    right,
    mobileProps,
    searchAriaLabel = 'Search',
    searchPlaceholder,
    searchTheme,
    searchValue,
    withBorder = true,
    ...headerProps
  } = props;
  const defaultMobileProps = {
    children,
    left,
    right,
    ...mobileProps,
  };

  const [searchOpened, setSearchOpened] = useState(false);
  const searchButtonRef = useClickOutside(() => setSearchOpened(false));

  const defaultTheme = useMantineTheme();
  const { classes } = useStyles();

  function handleClick(): void {
    setSearchOpened(!searchOpened);
  }

  const buttonClasses = [classes.button];
  if (searchOpened) {
    buttonClasses.push(classes.buttonOpened);
  }

  return (
    <>
      {/* Desktop Header */}
      <div className={hasResponsiveMode ? classes.sizeDesktop : undefined}>
        <MantineHeader
          height={searchOpened ? height + 110 : height}
          withBorder={withBorder}
          {...headerProps}
        >
          <Flex className={classes.container} h={height}>
            <Flex className={classes.around}>{left}</Flex>
            <Flex
              className={classes.menu}
              // @ts-expect-error wrong type for polymorphic component
              component={childrenComponent}
            >
              {children}
            </Flex>
            <Flex className={classes.around}>
              <div ref={searchButtonRef}>
                {Boolean(hasSearch) && (
                  <Button
                    aria-label={searchAriaLabel}
                    className={buttonClasses.join(' ')}
                    data-testid="search"
                    h={height}
                    onClick={handleClick}
                    variant="white"
                    w={height}
                  >
                    <MagnifyingGlass size={32} />
                  </Button>
                )}
                {Boolean(searchOpened) && (
                  <MantineProvider theme={searchTheme ?? defaultTheme}>
                    <HeaderSearch
                      data-testid="searchBar"
                      inputAriaLabel={searchAriaLabel}
                      inputPlaceholder={searchPlaceholder}
                      onChange={onSearchChange}
                      onSearchSubmit={onSearchSubmit}
                      opened={searchOpened}
                      value={searchValue}
                    />
                  </MantineProvider>
                )}
              </div>
              {right}
            </Flex>
          </Flex>
        </MantineHeader>
      </div>
      {/* Mobile Header */}
      <div className={hasResponsiveMode ? classes.sizeMobile : classes.none}>
        <HeaderMobile
          {...defaultMobileProps}
          hasSearch={hasSearch}
          onSearchChange={onSearchChange}
          onSearchSubmit={onSearchSubmit}
          searchInputProps={{
            placeholder: searchPlaceholder,
            title: searchAriaLabel,
          }}
          searchValue={searchValue}
          withBorder={withBorder}
          {...headerProps}
        />
      </div>
    </>
  );
}
