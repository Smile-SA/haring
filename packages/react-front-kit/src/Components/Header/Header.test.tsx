import { renderWithProviders } from '@smile/react-front-kit-shared/test-utils';

import { Header } from './Header';

describe('Header', () => {
  it('matches snapshot', () => {
    const { container } = renderWithProviders(
      <Header>
        <a href="#">Espace documentaire</a>
        <a href="#">Espace workflow</a>
        <a href="#">Archives</a>
      </Header>,
    );
    expect(container).toMatchSnapshot();
  });
});
