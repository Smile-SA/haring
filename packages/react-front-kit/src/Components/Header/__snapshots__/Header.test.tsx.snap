// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header matches snapshot 1`] = `
<div>
  <header
    class="mantine-Header-root mantine-1qb6iub"
  >
    <div
      class="mantine-o19s04"
    >
      <div
        class="mantine-1dcfulj"
      />
      <div
        class="mantine-jlqgta"
      >
        <a
          href="#"
        >
          Espace documentaire
        </a>
        <a
          href="#"
        >
          Espace workflow
        </a>
        <a
          href="#"
        >
          Archives
        </a>
      </div>
      <div
        class="mantine-1dcfulj"
      >
        <button
          aria-label="Search"
          class="mantine-UnstyledButton-root mantine-Button-root mantine-1ckwbd"
          data-button="true"
          data-testid="search"
          type="button"
        >
          <div
            class="mantine-1wpc1xj mantine-Button-inner"
          >
            <span
              class="mantine-1ryt1ht mantine-Button-label"
            >
              <svg
                fill="currentColor"
                height="32"
                viewBox="0 0 256 256"
                width="32"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M229.66,218.34l-50.07-50.06a88.11,88.11,0,1,0-11.31,11.31l50.06,50.07a8,8,0,0,0,11.32-11.32ZM40,112a72,72,0,1,1,72,72A72.08,72.08,0,0,1,40,112Z"
                />
              </svg>
            </span>
          </div>
        </button>
      </div>
    </div>
  </header>
</div>
`;
